components/brave_sync/brave_sync_service_impl.cc

```diff
@@ -763,35 +566,145 @@ void BraveSyncServiceImpl::OnResolvedPreferences(const RecordsList& records,
+    node->GetMetaInfo("object_id", &node_object_id);
+
+    if (node_object_id.empty())
+      continue;
+
+    if (object_id == node_object_id)
+      return node;
+  }
+  return nullptr;
+}
+
+uint64_t GetIndex(const bookmarks::BookmarkNode* root_node,
+                  const jslib::Bookmark& record) {
+  uint64_t index = 0;
+  ui::TreeNodeIterator<const bookmarks::BookmarkNode> iterator(root_node);
+  while (iterator.has_next()) {
+    const bookmarks::BookmarkNode* node = iterator.Next();
+    std::string node_order;
+    node->GetMetaInfo("order", &node_order);
+
+    if (!node_order.empty() &&
+        brave_sync::CompareOrder(record.order, node_order))
+      return index;
+
+    ++index;
+  }
+  return index;
+}
 
-  for (const auto &sync_record : records) {
+void UpdateNode(bookmarks::BookmarkModel* model,
+                const bookmarks::BookmarkNode* node,
+                const jslib::SyncRecord* record) {
+  auto bookmark = record->GetBookmark();
+  if (bookmark.isFolder) {
+    // SetDateFolderModified
+  } else {
+    model->SetURL(node, GURL(bookmark.site.location));
+    // TODO, AB: apply these:
+    // sync_bookmark.site.customTitle
+    // sync_bookmark.site.lastAccessedTime
+    // sync_bookmark.site.favicon
+  }
+  model->SetTitle(node,
+      base::UTF8ToUTF16(bookmark.site.title));
+  model->SetDateAdded(node, bookmark.site.creationTime);
+  model->SetNodeMetaInfo(node, "object_id", record->objectId);
+  model->SetNodeMetaInfo(node, "order", bookmark.order);
+}
+
+void BraveSyncServiceImpl::OnResolvedBookmarks(const RecordsList &records) {
+  bookmark_model_->BeginExtensiveChanges();
+  for (const auto& sync_record : records) {
     DCHECK(sync_record->has_bookmark());
-    LOG(ERROR) << "TAGAB brave_sync::BraveSyncServiceImpl::OnResolvedBookmarks: title=<" << sync_record->GetBookmark().site.title << ">";
-    LOG(ERROR) << "TAGAB brave_sync::BraveSyncServiceImpl::OnResolvedBookmarks: sync_record->objectId=<" << sync_record->objectId << ">";
-    LOG(ERROR) << "TAGAB brave_sync::BraveSyncServiceImpl::OnResolvedBookmarks: order=<" << sync_record->GetBookmark().order << ">";
     DCHECK(!sync_record->objectId.empty());
-    std::string local_id = sync_obj_map_->GetLocalIdByObjectId(storage::ObjectMap::Type::Bookmark, sync_record->objectId);
-    LOG(ERROR) << "TAGAB brave_sync::BraveSyncServiceImpl::OnResolvedBookmarks: local_id=<" << local_id << ">";
-
-    if (sync_record->action == jslib::SyncRecord::Action::CREATE && local_id.empty()) {
-      bookmarks_->AddBookmark(*sync_record);
-    } else if (sync_record->action == jslib::SyncRecord::Action::DELETE && !local_id.empty()) {
-      bookmarks_->DeleteBookmark(*sync_record);
-    } else if (sync_record->action == jslib::SyncRecord::Action::UPDATE && !local_id.empty()) {
-      bookmarks_->UpdateBookmark(*sync_record);
+
+    auto bookmark_record = sync_record->GetBookmark();
+    std::vector<const bookmarks::BookmarkNode*> nodes;
+    bookmark_model_->GetNodesByURL(
```

Comment:
@bridiver  is that for speedup the search?

Reply:
yes, since we know the url we might as well limit the list we have to iterate through